File: LeetCodeJS/courseSchedule.js
// courseSchedule.js
class CourseSchedule {
    static UNVISITED = 0;
    static VISITING = 1;
    static VISITED = 2;

    canFinish(numCourses, prerequisites) {
        const graph = new Map();

        // Build the graph
        prerequisites.forEach(([course, prereq]) => {
            if (!graph.has(course)) {
                graph.set(course, []);
            }
            graph.get(course).push(prereq);
        });

        const states = Array(numCourses).fill(CourseSchedule.UNVISITED);

        const dfs = (course) => {
            if (states[course] === CourseSchedule.VISITING) return false; // Cycle detected
            if (states[course] === CourseSchedule.VISITED) return true;   // Already visited

            states[course] = CourseSchedule.VISITING;
            const neighbors = graph.get(course) || [];

            for (let neighbor of neighbors) {
                if (!dfs(neighbor)) {
                    return false;
                }
            }

            states[course] = CourseSchedule.VISITED;
            return true;
        };

        for (let i = 0; i < numCourses; i++) {
            if (!dfs(i)) {
                return false;
            }
        }

        return true;
    }
}

// Export the CourseSchedule class
module.exports = CourseSchedule;

File: LeetCodePy/CourseSchedule.py
class CourseSchedule:
    def canFinish(self, numCourses: int, prerequisites: list[list[int]]) -> bool:
        graph = {i: [] for i in range(numCourses)}  # Create a graph
        for course, prereq in prerequisites:
            graph[course].append(prereq)
        
        # State tracking for each course: 0 = UNVISITED, 1 = VISITING, 2 = VISITED
        UNVISITED, VISITING, VISITED = 0, 1, 2
        states = [UNVISITED] * numCourses

        def dfs(course):
            if states[course] == VISITING:
                return False  # Cycle detected
            if states[course] == VISITED:
                return True  # Already visited
            
            # Mark the course as visiting
            states[course] = VISITING
            
            # Check all the neighbors (prerequisites)
            for prereq in graph[course]:
                if not dfs(prereq):
                    return False
            
            # Mark the course as fully visited
            states[course] = VISITED
            return True

        # Check if all courses can be finished
        for course in range(numCourses):
            if not dfs(course):
                return False
        return True
